dokka-subprojects/analysis-kotlin-api/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/sample/SampleRewriter.kt

```diff
@@ -0,0 +1,31 @@
+/*
+ * Copyright 2014-2024 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package org.jetbrains.dokka.analysis.kotlin.sample
+
+public typealias ShortFunctionName = String
+
+public interface FunctionCallRewriter{
+    /**
+     * In the case of a [trailing lambda](https://kotlinlang.org/docs/lambdas.html#passing-trailing-lambdas),
+     * it will be passed as it in the last element of [argumentList]
+     * E.g., the snippet `f(0) { i++ }` will have `"0"` and `"{ i++ }"` arguments
+     */
+    public fun rewrite(argumentList: List<String>, typeArgumentList: List<String>): String?
+}
+
+/**
+ * For simplicity's sake, it process sample code syntactically and does not have the information about semantics of code, e.g. resolved types.
+ *
+ *
+ * TODO write KDoc
+ */
+public interface SampleRewriter {
+    /**
+     * `null` to delete a whole import directive
+     */
+    public fun rewriteImportDirective(importPath: String): String? = importPath
+    public val functionCallRewriters: Map<ShortFunctionName, FunctionCallRewriter>
```

Comment:
What user case does using a function instead of map cover? Or what essential advantages of using it?
As for me, it can be a little confusing to have a function that returns `null` while `FunctionCallRewritter` can also return `null`.

A map can be deprecated easier than a function. For example, let's suppose for merging of multiple rewriters we need to know all rewriters, so a map could be useful.

Reply:
I agree with Oleg that having a function might be better than a map, and I have a slightly different suggestion to simplify it further: get rid of the `FunctionCallRewriter` interface at all, and move the rewrite function down into `SampleRewriter`:

```kotlin
public interface SampleRewriter {
    public fun rewriteFunctionCall(shortName: String, arguments: List<String>, typeArguments: List<String>): String? = null
}
```

The advantage here is that the user doesn't need to implement the interface for every single function name, and it's generally easier to understand how to use it and what's going on. It should also be easier to implement it (can be a single switch statement)

The disadvantage is that we'll be creating more garbage (creating lists for every function), but it doesn't affect 99.9% of projects that don't have a `SampleRewriter`. It also shouldn't be noticeable as there are bigger performance problems, and it can be optimized later if needed

The meaning of `null` changes here. Instead of meaning "remove this call", it'll mean "do not re-write this call", and if you want to remove it - I guess you'd return `""`, which looks more or less ok

What do you think?

Reply:
I think a such function is also ok. Additionally I like it is consistent with `rewriteImportDirective`.  
Originally the map was needed because of "do not re-write this call" and an empty string was reserved for keeping formatting.  
There is a 'problem' that a such function encourages to rewrite "do not re-write this call" to something like `$shortName(${arguments.joinToString()})` that can break formatting or even representation (in case of a lamda). 
I have no preference.

Reply:
While I do like the consistency with `rewriteImportDirective`, I don't really like, that we introduce such possible performance problem, as the amount of declarations we need to rewrite is small comparing to those we don't need to rewrite.
On current moment if we will go with just `rewriteFunctionCall` we will:
1. create a visitor for each function call
2. execute it to collect all arguments and type arguments of a function call
  * if we need to rewrite the call - we will create some string in rewriter
  * if not - we will still need to create a string or manually (via `$shortName(${arguments.joinToString()})` + somehow handle lambdas) or provide some unobvious `stub` return value (`""`) to not rewrite the call and also additionally construct string back.
So I think In this case we will only complicate things more for this niche use-case.

Regarding `Map` vs just function which return rewriter: overall the difference is minimal here.
But I think it's better to try to keep API smaller + with property and Map migration will be harder. If we will want to change something, as functions could have overloads. Or we will be able to introduce separate function with different name. While with a `Map` property it's a little harder to do this IMO.

> it can be a little confusing to have a function that returns null while FunctionCallRewritter can also return null.

As for me it's look logical, as we have 3 possible cases for a `function` call:
1. do nothing (`functionCallRewriter` returns `null`)
2. remove function (`FunctionCallRewritter` returns `null`, same as for `rewriteImportDirective`)
3. rewrite function (`FunctionCallRewritter` returns something)

> For example, let's suppose for merging of multiple rewriters we need to know all rewriters, so a map could be useful.

I don't see how this will help here, but may be :) Though, we don't have this requirement right now and also don't have an idea on how this should look, how to resolve conflicts, what about ordering, what about rewriting after rewriting and so on. So I would think, that if we will want to support this, there will be more problems and even `Map` could not be enough.

In the end, I don't have big preference over one or another API, as this API for now should not be used heavily :)

Reply:
To sum up, nobody does not mind a function instead of a map.
We can choose `public fun functionCallRewriter(name: String): FunctionCallRewritter?` since it does not have potential problems with performance and an ambiguous default value.
@IgnatBeresnev any objections/preferences?

Reply:
Sounds good to me